### ISSUE TITLE
Trace file is left when the process crashed before stopping the profiling

### ISSUE THREAD

**Issue by xiaomi7732** [microsoft/ApplicationInsights-Profiler-AspNetCore#204]
**Describe the bug**
RT. Basically, when the process is terminated unexpectedly, the temp profiling file will be left there.

**To Reproduce**
1. Start profiler.
2. Kill the process in the middle.
3. Check the tracing file.

**Expect**
temp file be deleted.

**Actual**
The temp file is still there.

**Comment by xiaomi7732** [comment]
Since the process was killed, there's little we can do during the session. However, what we can do, is to clean the trace files up in the next session, should the process started again.

The feature is built into the new beta release, no user action is needed to leverage it:

https://www.nuget.org/packages/Microsoft.ApplicationInsights.Profiler.AspNetCore/2.6.0-beta2

At the same time, it is also possible to customize the behavior:

```json
{
    "ServiceProfiler": {
        "TraceScavenger": {
            "InitialDelay": "00:00:00",    // Delay 0 seconds, no delay effectively.
            "Interval": "01:00:00",          // Run trace scavenger every 1 hour.
            "GracePeriod": "00:10:00"   // Do not delete the file if it was accessed within 10 minuts.
    }
}
```

**Comment by esbenbach** [comment]
Nice improvement, looking forward to it...

**Comment by esbenbach** [comment]
Not sure we are going to test it with the beta. Regardless we are not going
to release before january some time.
Ill look into using the beta package for that relase though.

One thing im almost certain about is that this is only half a solution.
There is still something causing the profiler to get stuck in the first
place, generating huge files.

But one thing at a time

**Comment by xiaomi7732** [comment]
2.6.0 is released. Please check out the stable package:
https://www.nuget.org/packages/Microsoft.ApplicationInsights.Profiler.AspNetCore/2.6.0

**Comment by esbenbach** [comment]
@xiami7732   We finally had the time to run this in production a bit.

The file is now removed, but it is still generated in the first place. So the issue is NOT solved, as a crash needs to occur before the file is cleaned up (and stops growing).

Secondly, all subsequent profiler writes (different processes on the same VM), will fail with "not enough disk space", even AFTER the file is removed and there is enough disk space.

**Comment by xiaomi7732** [comment]
@esbenbach Thanks for the confirming. There are several things that we need to touch base:

1. The file is generated - yes, that's by design as a cache to holding all events happened during the applications' running; I am not sure holding all those envets in the memory is a good idea. But the deal is, after the running, if the app does NOT crash, it should clean after itself. And it crashes, the file scavenger will do the clean up upon next start. Am I missing something?

2. A crash needs to occur before the file is cleaned up - that's not expected. The file scavenger should have been keep running. What happens if you set the running of it to more frequent:
```json
{
    "ServiceProfiler": {
        "TraceScavenger": {
            "InitialDelay": "00:00:00",    // Delay 0 seconds, no delay effectively.
            "Interval": "00:10:00",          // Run trace scavenger every 10 minutes.
            "GracePeriod": "00:5:00"   // Do not delete the file if it was accessed within 5 minuts.
    }
}
```
On the other hand, if there's no crash, files would have been removed automatically.

3. If there's enough space, it would be weird that it reports back as not enough space. Do you mind post a callstack for when the issue happens?

4. There's no other reports on this issue, which leads me to think, there's some special setup / environment that maybe I missed. Can you think of any?

**Comment by xiaomi7732** [comment]
Thanks for the response. The callstack is also interesting. Based on those, I have spotted something that are diff from the expectations.

1. Why file that worth of gbs - that is way larger than the usual trace file
    * The netperf trace is relatively low profile, a trace file is usually around the size of 50MB.
    * We have a 200MB circular buffer size by default - I assume you didn't overwrite the value to make it larger, could you please confirm?
    * I want to see the trace files to understand why. Could you please send me your application insights instrumentation key over the email (saars@microsoft.com)? With that, and your permission, I shall be able to check 1 piece of the netperf file to see more details about it.

2. From the callstack, I see an IO exception, that is not necessary out of space exception unless checked by the error code, right? Did you do something to confirm it?
  * If it is out of space, and you still see free space, is it possible the %UserProfile%/temp is mounted somewhere that you have less space than the other folder?
  * Another possibility is that there's no permission to write the trace file to the target folder (from the buffer) - that happens on a read-only container, but also could happen anywhere when writing is not allowed.
  * An easy way to rule those out is to set the local cache folder to a different place:
     For example: ServiceProfiler__LocalCacheFolder = "c:\TraceCache"
    * The default is set to: %UserProfile%\temp\ on Windows, /tmp/ on Linux.
    * See [here](https://github.com/microsoft/ApplicationInsights-Profiler-AspNetCore/blob/main/Configurations.md) for more user configurations.

**Comment by esbenbach** [comment]
1. No overwrite that i know of (I.e. if so its NOT done on purpose as I can't find any sort of setting for it).
Ill contact you on email for the instrumentation key.
2. The io exception message text says this (i did NOT check the exception code) : 
**There is not enough space on the disk. :** 'C:\SvcFab\_App\Infosoft.S4.ConfigurationService.FabricType_App117\temp\ASPNETCORE_83e1964b-b195-4fb4-bc34-2cc49d5ae882.tmp'

There is nothing "easy" about doing cache path adjustments in an SF cluster with both the "sdk" and the "sf" profiler running at the same time :) But it points to the same place....

There are NO containers in play here btw, only SF services.....(which in this  case hosts the kestrel process)

**Comment by xiaomi7732** [comment]
Hey @esbenbach,
Thanks for giving the details. Based on the info for now, I don't think we are looking into the right direction. **We need to figure out where does those huge file come from.** Here's the details:

I spent some time digging into the backend by your iKey and here are some discoveries:

1. In the last 30 days, we don't have telemetries of [Microsoft.ApplicationInsights.Profiler.AspNetCore](https://www.nuget.org/packages/Microsoft.ApplicationInsights.Profiler.AspNetCore) uploading profiler traces (I'll call it eventpipe profiler since the NuGet package is built upon the tech of EventPipe).
    1. To be clear, I DO see telemetries for that profiler to contact our backend, like request configurations, but there's no trace upload.
1. There are trace files, they are uploaded by the traditional profiler.
    1. The package size is around 50MB compressed, extracted to about 500MB ish
    2. Still off from what you are seeing - several gb.
    3. And this Profiler will not writing multiple files at a time, it is out of proc, and never reported fill up the disk space.
    4. I assume that is the SF Profiler that you mentioned. Am I correct to assume, this is how you turned on the SF Profiler: <https://learn.microsoft.com/en-us/azure/azure-monitor/profiler/profiler-servicefabric>?
1. In the trace files, I see 30+ processes of your services, do you know are they all have the EventPipe profiler turned on? Or is it just a few? There's a possibility that 30+ more want to writing at the same time and none got through. Let's see if that's the case or not.
1. I assume where you see the message "There is not enough space on the disk" and the callstack posted earlier doesn't come from the same exception, could you please confirm?
    1. The callstack provided pointing to the EventPipe Profiler. But the event pipe profiler doesn't write temp file prefix with "ASPNETCORE_", I am wondering where does that file come from.
    1. I will need to double check the behavior of the SF profiler but I suspect its not using that either.
    2. Do you happen to know where does that come from?

Basically, we need to find out where are those large gbs file come from. Otherwise, we might not fix the right product.

**Comment by esbenbach** [comment]
1. Not sure why that is, it should be enabled. I sent you the original Azure support request id that includes the code we use to configure the eventpipe profiler, but I can repost it here if you prefer (or send it via email directly).

I presume there are no uploads of the sizes because the application generating the large file is somehow busy generating the file (or stuck) and thus cannot upload it.

2. Your assumption is correct, that is how the profiler is enabled.

The snippet from our bicep template:
```bicep
     SinksConfig: {
                    Sink: [
                      {
                        name: 'ApplicationInsights'
                        ApplicationInsights: applicationInsightsKey
                      }
                      {
                        name: 'ApplicationInsightsProfilerSink'
                        ApplicationInsightsProfiler: applicationInsightsKey
                      }
                    ]
                  }
```

3. There are 25 asp.net core based "application services", most of those have 2 instances/replicas, so somewhere between 40 and 55 services, distributed across 9 VMs.

4. The call stack and the "There is not enough space on the disk" are NOT from the same exception. Because I took the wrong message, but it was from a similar error.
The message that belongs to the call stack above is 

`There is not enough space on the disk. : 'C:\SvcFab\_App\Infosoft.S4.Subscriber.ServiceFabricType_App108\temp\SPTraces\6bb4eba0-9694-40bc-84cc-7e273add4c1a.netperf'`

The last time it happend was on 2024-06-24 06:46:29 UTC    - some time prior to this another process has generated the large netperf file filling up disk space....

I am 99% certain this is another issue, provoked by the lack of disk space.
  i. Application A generates large netperf file (fill out disk space) - the core issue.
  ii: Application B need to create its own netperf file, has no disk space - creates exception.
  iii: Scavenger (or a human) removes the giant netperf file, free up space.
  iiii: Application B is either "stuck" trying to write netperf file, or needs to create a new netperf file and never checks for disk space again and just creates the exception.




Totally agree with the figuring out the large files (as mentioned in my inital comment to this PR I was afraid it was not the right fix :)) 


A few extra notes that possibly drowned in the support request:
- We have toggled off threshold triggers in the App Insights portal for CPU/Memory to reduce the number of times the issue occurs.
- It occurs less frequently the "more" resources the VMs have. We used to run with around 1gb free memory (a little less), but have scaled up to avoid the issue further.
- We have mostly seen it in our production environment, MAYBE because our test environment is wiped each night. Maybe due to lower load (it does have significantly less resources though).

**Comment by xiaomi7732** [comment]
@esbenbach, okay, it is then obviously a out of disk space exception. And the **netperf** file was created by this eventpipe profiler. But that shouldn't be too big (< 200Mb).

I think it would be more valuable to find out what is generating that big file: `ASPNETCORE_83e1964b-b195-4fb4-bc34-2cc49d5ae882.tmp`.

* Is it possible to run the lock smith (PowerToys) on the box to figure out the handler?
    * https://learn.microsoft.com/en-us/windows/powertoys/file-locksmith

* Does it look similar to these issues: 
    * https://stackoverflow.com/questions/62910894/asp-net-core-writes-request-body-to-windows-temp-folder
    * https://github.com/dotnet/aspnetcore/issues/24122?
    * Is it possible to try to set an environment variable of `ASPNETCORE_TEMP` to relocate those temp file to somewhere else?

Thanks.

**Comment by xiaomi7732** [comment]
I'll close this due to no response. Please feel free to reopen this once there's time for following up.

**Comment by esbenbach** [comment]
@xiaomi7732 - sorry for the delayed response (summer vacations).

1. The issue IS the netperf file being too big, at least it was before the scavenging was introduced. In the referenced azure support case, there should be example files for you to look at (I can't give you new ones since the scanvenging now actually deletes the files).
2. Given that the tmp file is not the big file, i doubt it helps, also the "tmp" file message was just me copy/pasting the wrong exception in our insights collection. That said, IF I find a ASPNETCORE tmp file again I can try to see if i can process it using lock smith.

Regarding the ASPNETCORE_TEMP it may be possible, but again I doubt its going to solve anything. I will look into trying to change the TEMP location....

**Comment by xiaomi7732** [comment]
@esbenbach Welcome back.

Please try the settings of the environment variable and let me know. The deal is that that temp file could grow up to GBs.

At the same time, do you mind share the link to the Azure Support ticket that you mentioned? I am wondering, what is the size of the netperf you attached there?

Assuming there's no bug in the buffer settings, each `.netperf` file is less than 200MB, and say you have 50 apps, worse case scenario, all of them are taking profiler at the same time, that's 10GB of space. Would that 10G be the problem on your system?

I'll confirm if there's issue of the buffer settings on our end.

**Comment by esbenbach** [comment]
Forwarded the latest email from the support case.

10Gb is not an issue, anf as mentioned its the netperf files that were big.

I will attempt to assign the environment variable to store tmp files on the ephemeral disk, but note that the occurrence happen rarely (maybe once a month or less).

**Comment by xiaomi7732** [comment]
Thank you @esbenbach for forwarding me the email. I didn't realize that you have 50Gb ish **netperf** files. I will look into what could cause that - that's not suppose to happen at all.

At the same time, as a work around, if you have a disk that is large enough to host those files, try set this environment variable:

```shell
ServiceProfiler__LocalCacheFolder = "A PATH that is large enough to workaround disk space full issue"
# That would overwrite the default of %UserProfile%\temp\ to store the .netperf file.
```

**Comment by esbenbach** [comment]
I will attempt to set both aspnet temp and localcachefolder variables. As mentioned, not sure if its easily doable, nor do I have any idea when it may have an effect :)

**Comment by xiaomi7732** [comment]
@esbenbach, I have looked into the implementation and here's some updates.

1. It is possible to have **huge netperf** files - my assumption of that it should just below 200MBs was wrong.
  1. I can't repro to get a 5G file, but it is easy for me to get a file larger than 200MS (I actually got a 500MB ish trace file).
1. The way it could happen is when you have very quick / short request/response. That leads to huge number of events during the matter of 2 minutes profiling.
  1. And that is by design in the effort to capture as much data as possible for analyzing.

In such scenario, taking 2 minutes of profiling trace would cause the issue of large trace files.

To make it work, considering the request / response is already pretty quick, we have a level of controlling of the profiling duration in the Portal.

Could you please try check what you have there and **reduce the duration for the sampling to 30 seconds**?

Here's how to do it:

1. Open the performance blade of your Azure Portal
1. Click "Profiler" button:
  ![image](https://github.com/user-attachments/assets/ff2c6823-d653-4a9e-a287-89c4ea64904b)
1. Set the duration for sampling to 30 seconds:
  ![image](https://github.com/user-attachments/assets/ffd2f472-7c5b-4fb3-b33e-bdd04c05fe31)

I believe that will address your problem of having large netperf files. If there's still large than expected trace file, please reopen this ticket and I'll take another look.

A side note, it is unfortunately I didn't get a chance to see the netperf file you had at the beginning. I wish that we can have a copy of the netperf file that we can dig into.

**Comment by esbenbach** [comment]
Can i disable the scavenger? Then I am certain we can get you a file
somehow :-)

**Comment by esbenbach** [comment]
Also the sampling time is already at 30 seconds. My best guess is that the profiler is somehow stuck, so it doesnt get to complete during the assigned window.

As also mentioned in the support request, high memory pressure seems to trigger the issue more frequently.

**Comment by xiaomi7732** [comment]
@esbenbach Technically, you could set it that the initial delay to be big number like:

ServiceProfiler__TraceScavenger__InitialDelay = 3650:00:00:00; // 365days * 10.

Once there's the `netperf` file, it would tell us the start time and the end time of the tracing, we then would know if that is due to stuck or is it still 30 seconds in duration.

**Comment by esbenbach** [comment]
Let me try changing the scavenger delay and see what happens.

**Comment by xiaomi7732** [comment]
@esbenbach How the CPU utilization is of your application? Is that CPU intense? The reason I ask is that I step upon a situation lately where the app is CPU intense, and no code moving forward to stop the profiling and leading to large trace file. Do you feel that could be the same cause?

**Comment by esbenbach** [comment]
@xiaomi7732  Each process is not very CPU intensive, and even together (all instances on a given VM) they rarely exceed 30-40% CPU utilization. HOWEVER as mentioned previously a few times, the operating parameters are very pressed for memory. Each process in isolation is not the memory intensive, but they are packed closely together so we often reach 90% memory.

It should be noted that reducing the memory utilization together with reduced sampling has almost removed the issue. It does still happen, but a lot less frequently, and the scavenging usually picks up the trace files and removes them before we have an issue of it.

**Comment by xiaomi7732** [comment]
Thanks for the update. Looks like they are different causes...

**Comment by xiaomi7732** [comment]
@esbenbach Is it still happening for you? If you want, try it out on this package, the code change there may help with your case:
https://www.nuget.org/packages/Microsoft.ApplicationInsights.Profiler.AspNetCore/3.0.2-beta3

**Comment by esbenbach** [comment]
Still happening. Upgrading to 3.x switches to an OpenTelemetry shim, so all
our other dependcy tracking etc, needs a revamp. We are looking into it,
unfortunately it is a bit hard to prioritize....

Esben Bach

**Comment by xiaomi7732** [comment]
@esbenbach Thanks for the update. I am just wondering, what is your reason for upgrading to `3.x`? Instead of switch to the Azure Monitor OpenTelemetry distro: https://www.nuget.org/packages/Azure.Monitor.OpenTelemetry.AspNetCore?

Thanks!

**Comment by esbenbach** [comment]
Not sure, we might switch to the opentelemetry distro, but i have yet to
research what may be missing (profiler and quick pulse used to be a case,
but maybe no more)

Esben Bach

**Comment by xiaomi7732** [comment]
@esbenbach Thanks for getting back. Okay. It seems like you are doing tech investigation, just so that you know, the Azure Monitor OTel Profiler works with 3.x, see details here:
* https://github.com/microsoft/ApplicationInsights-Profiler-AspNetCore/issues/248#issuecomment-4130085847

Also, a bit far away but OTEL is coming up with its own profiling signals: https://opentelemetry.io/blog/2024/profiling/
